File: Lista/quest14/singleton.go
package main

import (
	"fmt"
	"sync"
)

type Configuracao struct {
	config map[string]string
}

var (
	instancia *Configuracao
	once      sync.Once
)

func GetInstancia() *Configuracao {
	once.Do(func() {
		instancia = &Configuracao{
			config: make(map[string]string),
		}
	})
	return instancia
}

func (c *Configuracao) SetConfig(chave, valor string) {
	c.config[chave] = valor
}

func (c *Configuracao) GetConfig(chave string) string {
	return c.config[chave]
}

func main() {
	config1 := GetInstancia()
	config1.SetConfig("usuario", "admin")

	config2 := GetInstancia()
	fmt.Println(config2.GetConfig("usuario")) // Saída: admin
	fmt.Println(config1 == config2)             // Saída: true
}

File: Lista/quest14/singleton.py
class Configuracao:
    _instancia = None

    def __new__(cls):
        if cls._instancia is None:
            cls._instancia = super(Configuracao, cls).__new__(cls)
            cls._instancia.inicializar()
        return cls._instancia

    def inicializar(self):
        self.config = {}

    def set_config(self, chave, valor):
        self.config[chave] = valor

    def get_config(self, chave):
        return self.config.get(chave, None)

if __name__ == "__main__":
    config1 = Configuracao()
    config1.set_config("usuario", "admin")

    config2 = Configuracao()
    print(config2.get_config("usuario"))  # Saída: admin
    print(config1 is config2)  # Saída: True
